src/CameraZoomPlugin.cc

```diff
@@ -352,31 +393,91 @@ void CameraZoomPlugin::PreUpdate(
+  if (this->impl->zoomChanged) {
+    // Only calculate goal once each time zoom is changed.
+    const auto requestedZoomCmd = this->impl->zoomCommand.load();
+    const auto clampedZoomCmd = std::clamp(requestedZoomCmd,
+      this->impl->minZoom, this->impl->maxZoom);
+    if (std::abs(requestedZoomCmd - clampedZoomCmd) >
+      std::numeric_limits<double>::epsilon())
+    {
+      gzwarn << "Requested zoom command of " << requestedZoomCmd
+        << " has been clamped to " << clampedZoomCmd << ".\n";
+    }
+    this->impl->goalHfov =  this->impl->curHfov / clampedZoomCmd;
+    this->impl->zoomChanged = false;
+  }
+
+  // Goal is achieved, nothing to update.
+  if (std::abs(this->impl->goalHfov - this->impl->curHfov) <
+    std::numeric_limits<double>::epsilon())
     return;
 
   // Update component.
   sdf::Sensor &sensor = comp->Data();
   sdf::Camera *cameraSdf = sensor.CameraSensor();
+  assert(cameraSdf);  // Halt on debug mode intentionally.
+  if (!cameraSdf) {
+    // In release mode, return before a nullptr dereference.
+    return;
+  }
+
+  const auto oldHfov = cameraSdf->HorizontalFov().Radian();
+
+  // This is the amount of time passed since the last update.
+  const auto dt = _info.dt;
+  // How many meters the focal length could change per update loop
+  const auto maxFocalLengthChange = this->impl->slewRate *
+    std::chrono::duration<double>(dt).count();
+  const auto curFocalLength = cameraSdf->LensFocalLength();
 
-  this->impl->zoom = std::max(std::min(this->impl->zoomCommand.load(),
-      this->impl->maxZoom), this->impl->minZoom);
-  math::Angle oldHfov = cameraSdf->HorizontalFov();
-  math::Angle newHfov = this->impl->hfov / this->impl->zoom;
+  // This value should be static every iteration.
+  const auto sensorWidth = SensorWidth(curFocalLength, oldHfov);
 
+  const auto goalFocalLength = FovToFocalLength(
+    sensorWidth, this->impl->goalHfov);
+  // How many meters the focal length should change this iteration
+  const auto deltaFL = std::min(maxFocalLengthChange,
+    std::abs(curFocalLength - goalFocalLength));
+
+
+
+  // Update rendering camera with the latest focal length.
+  double newHfov;
+  if (goalFocalLength > curFocalLength) {
+    const auto newFocalLength =  curFocalLength + deltaFL;
+    newHfov = FocalLengthToFov(sensorWidth, newFocalLength);
+
+  } else {
+    const auto newFocalLength =  curFocalLength  - deltaFL;
+    newHfov = FocalLengthToFov(sensorWidth, newFocalLength);
+  }
   cameraSdf->SetHorizontalFov(newHfov);
   _ecm.SetChanged(cameraEntity, components::Camera::typeId,
-        ComponentState::OneTimeChange);
+    ComponentState::OneTimeChange);
 
   // Update rendering camera.
   this->impl->camera->SetHFOV(newHfov);
+}
 
-  gzdbg << "CameraZoomPlugin:\n"
-        << "Zoom:     " << this->impl->zoom << "\n"
-        << "Old HFOV: " << oldHfov << " rad\n"
-        << "New HFOV: " << newHfov << " rad\n";
+//////////////////////////////////////////////////
+double FocalLengthToFov(
```

Comment:
`FocalLengthToFov` is declared in `CameraZoomPlugin::Impl` so should be defined as  `CameraZoomPlugin::Impl::FocalLengthToFov `. Move implementation to the section where the other functions in `CameraZoomPlugin::Impl` are defined. Same for other static functions.

Reply:
Thanks, the new scoping makes sense. These functions are really generic lens formulas that I would expect to be somewhere with openCV, but I couldn't find them.